python-lib/query_handlers/scoring_handler.py

```diff
@@ -16,21 +16,64 @@ class ScoringHandler(QueryHandler):
     LEFT_NORMED_COUNT_AS = "_left_normed_count"
     RIGHT_NORMED_COUNT_AS = "_right_normed_count"
     ROW_NUMBER_AS = "_row_number"
+    TIMESTAMP_FILTERED_ROW_NB = "_timestamp_filtered_row_nb"
 
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self.sample_keys = [self.dku_config.users_column_name, self.dku_config.items_column_name]
+        self.precomputation_columns = self.sample_keys
         self.use_explicit = bool(self.dku_config.ratings_column_name)
+        self.timestamp_filtering = bool(self.dku_config.timestamps_column_name)
+        if self.use_explicit:
+            logger.debug("Using explicit feedbacks")
+            self.precomputation_columns += [self.dku_config.ratings_column_name]
+
+    def _build_timestamp_filtered(self, select_from, select_from_as="_prepared_input_dataset"):
+        def _build_timestamp_filtered_row_number(select_from_inner, select_from_as_inner):
+            ts_row_numbers = SelectQuery()
+            ts_row_numbers.select_from(select_from_inner, alias=select_from_as_inner)
+
+            columns_to_select_inner = self.precomputation_columns + [self.dku_config.timestamps_column_name]
+            self._select_columns_list(ts_row_numbers, column_names=columns_to_select_inner, table_name=select_from_as_inner)
+
+            ts_row_number_expression = (
+                Expression()
+                    .rowNumber()
+                    .over(
+                    Window(
+                        partition_by=[Column(self.based_column, table_name=select_from_as_inner)],
+                        order_by=[
```

Comment:
we should also order by items (pivot_column) so that if 2 items have the same timestamp we always order them the same way (for reproducible results in the tests).

Reply:
78a7acf